internal_transfers/actions/src/simulate/enso.ts

```diff
@@ -0,0 +1,95 @@
+// EnsoFinance made a transaction simulator which can be found and run here:
+// https://github.com/EnsoFinance/transaction-simulator
+// One can follow the instructions there to run a simulation API.
+import axios from "axios";
+import {
+  SimulationData,
+  SimulationParams,
+  TransactionSimulator,
+} from "./interface";
+import { Log } from "@tenderly/actions";
+import { Trace } from "../models";
+import { ethDeltaFromTraces } from "../utils";
+
+export class EnsoSimulator implements TransactionSimulator {
+  url: string;
+  accessKey?: string;
+
+  constructor(url: string, accessKey?: string) {
+    this.url = url;
+    this.accessKey = accessKey;
+  }
+
+  async simulate(params: SimulationParams): Promise<SimulationData> {
+    const {
+      sender,
+      contractAddress,
+      callData,
+      blockNumber,
+      value = "0",
+    } = params;
+    const response = await axios.post(
+      this.url,
+      {
+        chainId: 1,
+        from: sender,
+        to: contractAddress,
+        data: callData,
+        gasLimit: 5000000,
+        blockNumber,
+        value,
+      },
+      {
+        headers: {
+          "X-API-KEY": this.accessKey,
+          "Content-Type": "application/json",
+        },
+      }
+    );
+    if (!isEnsoSimulationResponse(response.data)) {
+      throw Error(`Invalid Response ${JSON.stringify(response.data)}`);
+    }
+    return parseEnsoSimulation(response.data);
+  }
+}
+
+interface EnsoSimulationResponse {
+  simulationId: string;
+  gasUsed: number;
+  blockNumber: number;
+  success: boolean;
+  trace: Trace[];
+  logs: Log[];
+}
+
+export function isEnsoSimulationResponse(
+  value: any
+): value is EnsoSimulationResponse {
+  if (!value || typeof value !== "object") return false;
+  const { simulationId, gasUsed, blockNumber, success, trace, logs } = value;
+  if (!simulationId || typeof simulationId !== "number") return false;
+  if (!gasUsed || typeof gasUsed !== "number") return false;
+  if (!blockNumber || typeof blockNumber !== "number") return false;
+  if (!success || typeof success !== "boolean") return false;
+  return Array.isArray(logs) && Array.isArray(trace);
+  // Even more could be done here (like 0x prefixed strings, etc...)
+  // or logs & traces are actually logs and traces.
+}
```

Comment:
I'm not a huge fan of a hand-written half baked validation function. I think it's fine to assume that if the API responds with 200 the response is well formed (or if we are sceptical use some automated validation that's based on the EnsoSimulationResponse interface type and doesn't require us writing this by hand ourself)

Reply:
I agree its incredibly gross and also not very helpful for debugging (since when this fails I can never tell exactly what was wrong. However, this is what the TS folks said to do in these situations....  I believe this is what allows me to write the function signature 
```ts
function parseEnsoSimulation(simulation: EnsoSimulationResponse): SimulationData
```
and actually give it a type.